zebra-network/src/peer_set/initialize/recent_by_ip.rs

```diff
@@ -0,0 +1,144 @@
+//! A set of IPs from recent connection attempts.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    net::IpAddr,
+    time::{Duration, Instant},
+};
+
+use crate::constants;
+
+#[derive(Debug)]
+/// Stores IPs of recently attempted inbound connections.
+pub struct RecentByIp {
+    /// The list of IPs in increasing connection time order.
+    pub by_time: VecDeque<(IpAddr, Instant)>,
```

Comment:
Optional: If we use `BTreeMap`, it will automatically sort times in increasing order. We can also use `self.by_time.split_off(now + self.time_limit)` to find the expired entries, which is slightly more CPU efficient than popping and comparing each entry.

https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.split_off

```suggestion
    pub by_time: BTreeMap<Instant, IpAddr>,
```

Reply:
Good idea.

I kept the `VecDeque` because I wasn't sure how to make a `BTreeMap` work faster with a configurable `max_connections_per_ip`, but changed it to use `partition_point()` (binary search) and `split_off()` https://github.com/ZcashFoundation/zebra/pull/7041/commits/397e70391a5b80a0097b58f740fc1a9eb8c0e7a8.

Reply:
I opened PR #7214 as an example of this change, please don't feel like you have to merge it. PR #7213 also contains a necessary simplification for this change: moving the time calculation out of the partition closure.